src/SDDK/wave_functions.hpp

```diff
@@ -27,482 +27,1970 @@
+
+    /* rank 0 stores the G=0 component */
+    if (wf.comm().rank() != 0) {
+        return;
+    }
+
+    auto ld = wf.ld() * 2;
+
+    auto sp = wf.actual_spin_index(spins__.begin());
+
+    auto ptr = wf.at(mem__, 0, sp, wf::band_index(br__.begin()));
+    auto m = br__.size();
+
+    if (is_device_memory(mem__)) {
+#if defined(SIRIUS_GPU)
+        if (std::is_same<T, double>::value) {
+            accblas::dscal(m, reinterpret_cast<double*>(scale__), reinterpret_cast<double*>(ptr), ld);
+        } else if (std::is_same<T, float>::value) {
+            accblas::sscal(m, reinterpret_cast<float*>(scale__), reinterpret_cast<float*>(ptr), ld);
+        }
+#else
+        RTE_THROW("not compiled with GPU support!");
+#endif
+    } else {
+        if (std::is_same<T, double>::value) {
+            FORTRAN(dscal)(&m, reinterpret_cast<double*>(scale__), reinterpret_cast<double*>(ptr), &ld);
+        } else if (std::is_same<T, float>::value) {
+            FORTRAN(sscal)(&m, reinterpret_cast<float*>(scale__), reinterpret_cast<float*>(ptr), &ld);
+        }
+    }
+}
+
+/// Compute inner product between the two sets of wave-functions.
+/**
+ * \tparam T Precision type of the wave-functions (float or double).
+ * \tparam F Type of the subspace (float or double for Gamma-point calculation,
+ *           complex<float> or complex<double> otherwise).
+ *
+ * \param [in] spla_ctx   Context of the SPLA library.
+ * \param [in] mem        Location of the input wave-functions (host or device).
+ * \param [in] spins      Spin range of the wave-functions.
+ * \param [in] wf_i       Left hand side of <wf_i | wf_j> product.
+ * \param [in] br_i       Band range of the <wf_i| wave-functions.
+ * \param [in] wf_j       Right hand side of <wf_i | wf_j> product.
+ * \param [in] br_j       Band range of the |wf_j> wave-functions.
+ * \param [out] result    Resulting inner product matrix.
+ * \param [in] irow0      Starting row of the output sub-block.
+ * \param [in] jcol0      Starting column of the output sub-block.
+ * \return None
+ *
+ * Depending on the spin range this functions computes the inner product between individaul spin components
+ * or between full spinor wave functions:
+ * \f[
+ *    M_{irow0+i,jcol0+j} = \sum_{\sigma=s0}^{s1} \langle \phi_{i0 + i}^{\sigma} | \phi_{j0 + j}^{\sigma} \rangle
+ * \f]
+ * where i0 and j0 and the dimensions of the resulting inner product matrix are determined by the band ranges for
+ * bra- and ket- states.
+ *
+ * The location of the wave-functions data is determined by the mem parameter. The result is always returned in the
+ * CPU memory. If resulting matrix is allocated on the GPU memory, the result is copied to GPU as well.
+ */
+template <typename T, typename F, typename W>
+inline std::enable_if_t<std::is_same<T, real_type<F>>::value, void>
+inner(::spla::Context& spla_ctx__, sddk::memory_t mem__, spin_range spins__, W const& wf_i__,
```

Comment:
Would be nice to generalize this to arbitrary arguments for beta-projectors and wave-functions. But we can also try to this after the PR is merged.

Reply:
Do you mean for any wave-function like objects (both <bra| and |ket>)?

Reply:
Yes, I think that would be the best, to have a template argument for both bra and ket.